resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesExternalShuffleManager.scala

```diff
@@ -0,0 +1,158 @@
+  }.getOrElse(Utils.getConfiguredLocalDirs(sparkConf))
+  private var shufflePodCache = scala.collection.mutable.Map[String, String]()
+  private var watcher: Watch = _
+
+  override def start(appId: String): Unit = {
+    // seed the initial cache.
+    val pods = client.pods()
+      .inNamespace(shuffleNamespace)
+      .withLabels(shufflePodLabels.asJava)
+      .list()
+    pods.getItems.asScala.foreach {
+      pod =>
+        if (Readiness.isReady(pod)) {
+          addShufflePodToCache(pod)
+        } else {
+          logWarning(s"Found unready shuffle pod ${pod.getMetadata.getName} " +
+            s"on node ${pod.getSpec.getNodeName}")
+        }
+    }
+
+    watcher = client
+      .pods()
+      .inNamespace(shuffleNamespace)
+      .withLabels(shufflePodLabels.asJava)
+      .watch(new Watcher[Pod] {
+        override def eventReceived(action: Watcher.Action, p: Pod): Unit = {
+          action match {
+            case Action.DELETED | Action.ERROR =>
+              shufflePodCache.remove(p.getSpec.getNodeName)
+            case Action.ADDED | Action.MODIFIED if Readiness.isReady(p) =>
+              addShufflePodToCache(p)
+          }
+        }
+        override def onClose(e: KubernetesClientException): Unit = {}
+      })
+    shuffleClient.init(appId)
+  }
+
+  private def addShufflePodToCache(pod: Pod): Unit = shufflePodCache.synchronized {
+    if (shufflePodCache.contains(pod.getSpec.getNodeName)) {
+      val registeredPodName = shufflePodCache.get(pod.getSpec.getNodeName).get
+      logError(s"Ambiguous specification of shuffle service pod. " +
+        s"Found multiple matching pods: ${pod.getMetadata.getName}, " +
+        s"${registeredPodName} on ${pod.getSpec.getNodeName}")
+
+      throw new SparkException(s"Ambiguous specification of shuffle service pod. " +
+        s"Found multiple matching pods: ${pod.getMetadata.getName}, " +
+        s"${registeredPodName} on ${pod.getSpec.getNodeName}")
+    } else {
+        shufflePodCache(pod.getSpec.getNodeName) = pod.getStatus.getPodIP
+    }
+  }
+
+  override def stop(): Unit = {
+    watcher.close()
+    shuffleClient.close()
+  }
+
+  override def getShuffleServiceConfigurationForExecutor(executorPod: Pod)
+      : Seq[(String, String)] = {
+    val nodeName = executorPod.getSpec.getNodeName
+    val shufflePodIp = shufflePodCache.synchronized {
+      shufflePodCache.get(nodeName).getOrElse(
+          throw new SparkException(s"Unable to find shuffle pod on node $nodeName"))
+    }
+    // Inform the shuffle pod about this application so it can watch.
+    shuffleClient.registerDriverWithShuffleService(shufflePodIp, externalShufflePort)
+    Seq((SPARK_SHUFFLE_SERVICE_HOST.key, shufflePodIp))
+  }
+
+  override def getExecutorShuffleDirVolumesWithMounts(): Seq[(Volume, VolumeMount)] = {
+    shuffleDirs.zipWithIndex.map {
+      case (shuffleDir, shuffleDirIndex) =>
+        val volumeName = s"$shuffleDirIndex-${FilenameUtils.getBaseName(shuffleDir)}"
```

Comment:
seems like this fixes a bug from before, where the old implementation would have attempted to mount multiple volumes with the same name?

Reply:
Yes.